File: geo3x3.lua
local geo3x3 = {}

geo3x3.encode = function(lat, lng, level)
  if level < 1 then
    return null
  end
  res = ""
  if lng >= 0.0 then
    res = "E"
  else
    res = "W"
    lng = lng + 180.0
  end
  lat = 90.0 - lat
  unit = 180.0

  for i = 1, level - 1 do
    unit = unit / 3
    x = math.floor(lng / unit)
    y = math.floor(lat / unit)
    res = res .. string.format("%d", x + y * 3 + 1)
    lng = lng - x * unit
    lat = lat - y * unit
  end
  return res
end

geo3x3.decode = function(code)
  c = string.sub(code, 1, 1)
  begin = 0
  flg = false
  if c == "-" or c == "W" then
    flg = true
    begin = 1
  elseif c == "+" or c == "E" then
    begin = 1
  end
  unit = 180.0
  lat = 0.0
  lng = 0.0
  level = 1
  for i = begin, string.len(code) - 1 do
    n = tonumber(string.sub(code, i + 1, i + 1))
    if n == 0 then
      break
    end
    unit = unit / 3
    n = n - 1
    lng = lng + math.floor(n % 3) * unit
    lat = lat + math.floor(n / 3) * unit
    level = level + 1
  end
  lat = lat + unit / 2
  lng = lng + unit / 2
  lat = 90.0 - lat
  if flg then
    lng = lng - 180.0
  end
  return {lat, lng, level, unit}
end

return geo3x3

File: geo3x3.vb
Public Class Geo3x3
    Public Shared Function Encode(lat As Double, lng As Double, level As Integer) As String
        If level < 1 Then
            Return ""
        End If
        Dim res As String = "E"
        If lng < 0.0 Then
            res = "W"
            lng += 180.0
        End If
        lat += 90.0
        Dim unit As Double = 180.0
        Dim i As Integer
        For i = 1 To level - 1
            unit /= 3.0
            Dim x As Integer = CInt(Math.Floor(lng / unit))
            Dim y As Integer = CInt(Math.Floor(lat / unit))
            Dim n As Integer = x + y * 3 + 1
            res = res & n.ToString
            lng -= x * unit
            lat -= y * unit
        Next
        Return res
    End Function

    Public Shared Function Decode(code As String) As Double()
        Dim res(4) As Double
        Dim clen As Integer = Len(code)
        If clen = 0 Then
            Return res
        End If
        Dim begin As Integer = 0
        Dim flg As Boolean = False
        Dim c As String = Mid(code, 1, 1)
        If c = "-" or c = "W" Then
            flg = True
            begin = 1
        Else If c = "+" or c = "E" Then
            begin = 1
        End If
        Dim unit As Double = 180.0
        Dim lat As Double = 0.0
        Dim lng As Double = 0.0
        Dim level As Integer = 1
        Dim i As Integer
        For i = begin To clen - 1
            Dim n As Integer
            n = InStr("0123456789", Mid(code, i + 1, 1)) - 1
            if n <= 0 Then
                Exit For
            End If
            unit /= 3.0
            n -= 1
            lng += (n mod 3) * unit
            lat += Math.Floor(n / 3) * unit
            level += 1
        Next
        lat += unit / 2.0
        lng += unit / 2.0
        lat -= 90.0
        If flg Then
            lng -= 180.0
        End If
        res(0) = lat
        res(1) = lng
        res(2) = level
        res(3) = unit
        Return res
    End Function
End Class
